ConnectButton Tests using Jest

shannonlal opened the pull request:
The following is a first stab at the unit testing the front end components using Jest.  The goal is to help document the front end and resolve some of the UI issues.  I am hoping by breaking this down into small Unit Tests it will make it easy to see the expected behaviour of each of the components.  I welcome any and all comments for this.

@n-riesco If you get a chance could you have a look at this.  Are you the best person to review this or should I run this by someone else?

n-riesco commented on app/components/Settings/ConnectButton/ConnectButton.react.js:
```diff
@@ -12,6 +22,38 @@ export default class ConnectButton extends Component {
+    * Will check whether connection requests are equal to or greater then 400
+    * @returns {boolean} true if connection error
+    */
+    isConnectionError() {
```
How about `hasConnectionFailed` or `connectionFailed`?

n-riesco commented on app/components/Settings/ConnectButton/ConnectButton.react.js:
```diff
@@ -12,6 +22,38 @@ export default class ConnectButton extends Component {
+     * @returns {boolean} true if status is loading
+     */
+    loadingStatus() {
+        return (isLoading(this.props.connectRequest.status) || isLoading(this.props.saveConnectionsRequest.status));
```
`isLoading` is nw only used here. I'd get rid of `isLoading` and rename `loadingStatus` to `isLoading` (it's confusing to have two function so similar in name and purpose)

n-riesco commented on app/components/Settings/ConnectButton/ConnectButton.react.js:
```diff
@@ -12,6 +22,38 @@ export default class ConnectButton extends Component {
+     * Checks whether the connection request status (HTTP) is greater or equal to 200 and less then 300
+     * @returns {boolean} true if connection request status >=200 and <300
+     */
+    isValidConnection() {
```
How about `isConnected`?

n-riesco commented on app/components/Settings/ConnectButton/ConnectButton.react.js:
```diff
@@ -12,6 +22,38 @@ export default class ConnectButton extends Component {
+     * Checks whether connection status is defined
+     * @returns {boolean} true if connection status is invalid
+     */
+    isInvalidConnectionStatus() {
```
This one if weird. The function name  implies that it checks whether the connection failed, but the code checks we're still waiting for a response to out connection request.

I also suspect this may be the source of bug #372 (see my comment below; I'll test if I'm right).

n-riesco commented on package.json:
```diff
@@ -212,7 +213,14 @@
     "style-loader": "^0.19.0",
     "webpack": "^3.8.1",
-    "yamljs": "^0.3.0"
+    "yamljs": "^0.3.0",
```
we keep the dependencies in alphabetical order

n-riesco commented on test/app/components/Settings/ConnectButton/ConnectButton.test.js:
```diff
@@ -0,0 +1,289 @@
+jest.dontMock('../../../../../app/components/Settings/ConnectButton/ConnectButton.react.js');
```
Are we affected by [this issue](https://facebook.github.io/jest/docs/en/troubleshooting.html#i-m-using-babel-and-my-unmocked-imports-aren-t-working) with unmocked imports?

shannonlal replied:
No.  Jest mocks modules by default.  I specify that it should not mock this component so that we can test the component.  This is a pattern that I have used before on other projects and seems to have worked

n-riesco replied:
I understand. What I meant is whether we need to use `unmock` instead.

shannonlal replied:
@n-riesco It looks like unmock is now the recommended approach.  Here is a link to a link on SO (https://stackoverflow.com/questions/36571357/difference-between-unmock-and-dontmock-in-jest).  I have updated the test and and added it to the PR

n-riesco commented on package.json:
```diff
@@ -212,7 +213,14 @@
+    "jest":"^22.3.0",
+    "babel-jest":"latest",
+    "react-test-renderer":"latest",
+    "react-addons-test-utils": "latest",
```
I think we don't need `react-addons-test-utils`, as we're using React 15.5

n-riesco commented on package.json:
```diff
@@ -212,7 +213,14 @@
-    "yamljs": "^0.3.0"
+    "yamljs": "^0.3.0",
+    "jest":"^22.3.0",
+    "babel-jest":"latest",
```
pin to major version number

n-riesco commented on test/app/components/Settings/ConnectButton/ConnectButton.test.js:
```diff
@@ -0,0 +1,289 @@
+describe('Connect Button Tests', () => {
+
+    beforeAll(() => {
+        // Setup the test
```
✂️

n-riesco commented on test/app/components/Settings/ConnectButton/ConnectButton.test.js:
```diff
@@ -0,0 +1,289 @@
+import Adapter from 'enzyme-adapter-react-15';
+
+describe('Connect Button Tests', () => {
+
```
~~Tests~~

n-riesco commented on test/app/components/Settings/ConnectButton/ConnectButton.test.js:
```diff
@@ -0,0 +1,289 @@
+    });
+
+    it('Should verify Connection Request Error', () => {
+        expect(ConnectButton).toBeDefined();
```
no need for this check (import would've failed before)

n-riesco commented on test/app/components/Settings/ConnectButton/ConnectButton.test.js:
```diff
@@ -0,0 +1,289 @@
+            editMode={editMode}
+        />);
+
+        expect(button).toBeDefined();
```
no need for this test

n-riesco commented on test/app/components/Settings/ConnectButton/ConnectButton.test.js:
```diff
@@ -0,0 +1,289 @@
+        />);
+
+        expect(button).toBeDefined();
+        expect(button.instance().isConnectionError()).toBeTruthy();
```
I'd tighten this test: `.toBe(true)`

n-riesco commented on test/app/components/Settings/ConnectButton/ConnectButton.test.js:
```diff
@@ -0,0 +1,289 @@
+    });
+
+    it('Should verify Save Connections Request Error', () => {
+        expect(ConnectButton).toBeDefined();
```
✂️

n-riesco commented on test/app/components/Settings/ConnectButton/ConnectButton.test.js:
```diff
@@ -0,0 +1,289 @@
+            editMode={editMode}
+        />);
+
+        expect(button).toBeDefined();
```
✂️

n-riesco commented on test/app/components/Settings/ConnectButton/ConnectButton.test.js:
```diff
@@ -0,0 +1,289 @@
+        />);
+
+        expect(button).toBeDefined();
+        expect(button.instance().isConnectionError()).toBeTruthy();
```
toBe(true)

n-riesco commented on test/app/components/Settings/ConnectButton/ConnectButton.test.js:
```diff
@@ -0,0 +1,289 @@
+    });
+
+    it('Should verify Save Connections Request without Error', () => {
+        expect(ConnectButton).toBeDefined();
```
✂️

n-riesco commented on test/app/components/Settings/ConnectButton/ConnectButton.test.js:
```diff
@@ -0,0 +1,289 @@
+            editMode={editMode}
+        />);
+
+        expect(button).toBeDefined();
```
✂️

n-riesco commented on test/app/components/Settings/ConnectButton/ConnectButton.test.js:
```diff
@@ -0,0 +1,289 @@
+        />);
+
+        expect(button).toBeDefined();
+        expect(button.instance().isConnectionError()).toBeFalsy();
```
toBe(false)

n-riesco commented on test/app/components/Settings/ConnectButton/ConnectButton.test.js:
```diff
@@ -0,0 +1,289 @@
+    });
+
+    it('Should verify loading connection request loading status', () => {
+        expect(ConnectButton).toBeDefined();
```
✂️

n-riesco commented on test/app/components/Settings/ConnectButton/ConnectButton.test.js:
```diff
@@ -0,0 +1,289 @@
+            editMode={editMode}
+        />);
+
+        expect(button).toBeDefined();
```
✂️

n-riesco commented on test/app/components/Settings/ConnectButton/ConnectButton.test.js:
```diff
@@ -0,0 +1,289 @@
+        />);
+
+        expect(button).toBeDefined();
+        expect(button.instance().loadingStatus()).toBeTruthy();
```
toBe(true)

n-riesco commented on test/app/components/Settings/ConnectButton/ConnectButton.test.js:
```diff
@@ -0,0 +1,289 @@
+    });
+
+    it('Should verify loading save connections request loading status', () => {
+        expect(ConnectButton).toBeDefined();
```
✂️

n-riesco commented on test/app/components/Settings/ConnectButton/ConnectButton.test.js:
```diff
@@ -0,0 +1,289 @@
+            editMode={editMode}
+        />);
+
+        expect(button).toBeDefined();
```
✂️

n-riesco commented on test/app/components/Settings/ConnectButton/ConnectButton.test.js:
```diff
@@ -0,0 +1,289 @@
+        />);
+
+        expect(button).toBeDefined();
+        expect(button.instance().loadingStatus()).toBeTruthy();
```
toBe(true)

n-riesco commented on test/app/components/Settings/ConnectButton/ConnectButton.test.js:
```diff
@@ -0,0 +1,289 @@
+    });
+
+    it('Should verify not loading status', () => {
+        expect(ConnectButton).toBeDefined();
```
✂️

n-riesco commented on test/app/components/Settings/ConnectButton/ConnectButton.test.js:
```diff
@@ -0,0 +1,289 @@
+            editMode={editMode}
+        />);
+
+        expect(button).toBeDefined();
```
✂️

n-riesco commented on test/app/components/Settings/ConnectButton/ConnectButton.test.js:
```diff
@@ -0,0 +1,289 @@
+        />);
+
+        expect(button).toBeDefined();
+        expect(button.instance().loadingStatus()).toBeFalsy();
```
toBe(false)

n-riesco commented on test/app/components/Settings/ConnectButton/ConnectButton.test.js:
```diff
@@ -0,0 +1,289 @@
+    });
+
+    it('Should verify valid connection', () => {
+        expect(ConnectButton).toBeDefined();
```
✂️

n-riesco commented on test/app/components/Settings/ConnectButton/ConnectButton.test.js:
```diff
@@ -0,0 +1,289 @@
+            editMode={editMode}
+        />);
+
+        expect(button).toBeDefined();
```
✂️

n-riesco commented on test/app/components/Settings/ConnectButton/ConnectButton.test.js:
```diff
@@ -0,0 +1,289 @@
+
+        expect(button).toBeDefined();
+
+        expect(button.instance().isValidConnection()).toBeTruthy();
```
toBe(true)

n-riesco commented on test/app/components/Settings/ConnectButton/ConnectButton.test.js:
```diff
@@ -0,0 +1,289 @@
+        />);
+
+        expect(button).toBeDefined();
+        expect(button.instance().isValidConnection()).toBeFalsy();
```
toBe(false)

n-riesco commented on test/app/components/Settings/ConnectButton/ConnectButton.test.js:
```diff
@@ -0,0 +1,289 @@
+
+        expect(buttonWrapper).toBeDefined();
+        expect(buttonWrapper.instance().isValidConnection()).toBeFalsy();
+    });
```
toBe(false)

n-riesco commented on test/app/components/Settings/ConnectButton/ConnectButton.test.js:
```diff
@@ -0,0 +1,289 @@
+    });
+
+    it('Should verify valid connection', () => {
+        expect(ConnectButton).toBeDefined();
```
✂️

n-riesco commented on test/app/components/Settings/ConnectButton/ConnectButton.test.js:
```diff
@@ -0,0 +1,289 @@
+            editMode={editMode}
+        />);
+
+        expect(button).toBeDefined();
```
✂️

n-riesco commented on test/app/components/Settings/ConnectButton/ConnectButton.test.js:
```diff
@@ -0,0 +1,289 @@
+    });
+
+    it('Should verify connection status', () => {
+        expect(ConnectButton).toBeDefined();
```
✂️

n-riesco commented on test/app/components/Settings/ConnectButton/ConnectButton.test.js:
```diff
@@ -0,0 +1,289 @@
+            editMode={editMode}
+        />);
+
+        expect(buttonWrapper).toBeDefined();
```
✂️

n-riesco commented on test/app/components/Settings/ConnectButton/ConnectButton.test.js:
```diff
@@ -0,0 +1,289 @@
+    });
+
+    it('Should create valid connected button', () => {
+        expect(ConnectButton).toBeDefined();
```
✂️

n-riesco commented on test/app/components/Settings/ConnectButton/ConnectButton.test.js:
```diff
@@ -0,0 +1,289 @@
+            editMode={editMode}
+        />);
+
+        expect(button).toBeDefined();
```
✂️

n-riesco commented on test/app/components/Settings/ConnectButton/ConnectButton.test.js:
```diff
@@ -0,0 +1,289 @@
+    });
+
+    it('Should connected button with error due to connect request error', () => {
+        expect(ConnectButton).toBeDefined();
```
✂️

n-riesco commented on test/app/components/Settings/ConnectButton/ConnectButton.test.js:
```diff
@@ -0,0 +1,289 @@
+            editMode={editMode}
+        />);
+
+        expect(button).toBeDefined();
```
✂️

n-riesco commented on test/app/components/Settings/ConnectButton/ConnectButton.test.js:
```diff
@@ -0,0 +1,289 @@
+    });
+
+    it('Should connected button with error due to connect request error', () => {
+        expect(ConnectButton).toBeDefined();
```
✂️

n-riesco commented on test/app/components/Settings/ConnectButton/ConnectButton.test.js:
```diff
@@ -0,0 +1,289 @@
+            editMode={editMode}
+        />);
+
+        expect(button).toBeDefined();
```
✂️

n-riesco commented on app/components/Settings/ConnectButton/ConnectButton.react.js:
```diff
@@ -4,6 +4,16 @@ import {pathOr} from 'ramda';
+ * The following is the Connect Button which triggers the connection
+ * @param {function} connect - Connect function
+ * @param {object} connectRequest - Connection Request
+ * @param {number || string} connectRequest.status -- 400 or loading
```
Also `* @param {Error} [connectRequest.error]`

n-riesco commented on app/components/Settings/ConnectButton/ConnectButton.react.js:
```diff
@@ -4,6 +4,16 @@ import {pathOr} from 'ramda';
+ * @param {object} connectRequest - Connection Request
+ * @param {number || string} connectRequest.status -- 400 or loading
+ * @param {object} saveConnectionsRequest - Saved Connection Request
+ * @param {number || string } saveConnectionsRequest.status -- 400 or loading
```
Also `* @param {Error} [saveConnectionsRequest.error]`

n-riesco commented on app/components/Settings/ConnectButton/ConnectButton.react.js:
```diff
@@ -4,6 +4,16 @@ import {pathOr} from 'ramda';
 
 const isLoading = (status) => status === 'loading';
 
+/**
```
This JSDoc comment should go above `static propTypes`

n-riesco commented on app/components/Settings/ConnectButton/ConnectButton.react.js:
```diff
@@ -4,6 +4,16 @@ import {pathOr} from 'ramda';
+ * @param {object} saveConnectionsRequest - Saved Connection Request
+ * @param {number || string } saveConnectionsRequest.status -- 400 or loading
+ * @param {boolean} editMode  - Enabled if Editting credentials
+ * @returns {ConnectButton}
```
~~@returns {ConnectButton}~~

n-riesco commented on app/components/Settings/ConnectButton/ConnectButton.react.js:
```diff
@@ -39,22 +81,16 @@ export default class ConnectButton extends Component {
                 buttonText = 'Connected';
             }
-        } else if (!connectRequest.status) {
+        } else if (this.isInvalidConnectionStatus()) {
```
The `return` below isn't properly indented. Could you fix it, please?

n-riesco commented:
@shannonlal A very comprehensive set of specs!

I find the logic in `render()` difficult to follow. How about something like this?

```js
    /**
    * @returns {boolean} true if waiting for a response to a connection request
    */
    isConnecting() {
        return this.props.connectRequest.status === 'loading';
    }

    /**
    * @returns {boolean} true if successfully connected to database
    */
    isConnected() {
        const status = Number(this.props.connectRequest.status);
        return (status >= 200 || status < 300);
    }

    /**
    * @returns {boolean} true if successfully connected to database
    */
    connectionFailed() {
        return Number(this.props.connectRequest.status) >= 400;
    }

    /**
    * @returns {boolean} true if waiting for a response to a save request
    */
    isSaving() {
        return this.props.saveConnectionsRequest.status === 'loading';
    }

    /**
    * @returns {boolean} true if connection has been saved
    */
    isSaved() {
        const status = Number(this.props.saveConnectionsRequest.status);
        return (status >= 200 || status < 300);
    }

    /**
    * @returns {boolean} true if successfully connected to database
    */
    saveFailed() {
        return Number(this.props.saveConnectionsRequest.status) >= 400;
    }

    render() {
        const {
            connect,
            connectRequest,
            saveConnectionsRequest,
            editMode
        } = this.props;

        let buttonText;
        let buttonClick = () => {};
        let error = null;

        if (!editMode) {
            buttonText = 'Connected';

        } else if (this.isConnecting() || this.isSaving()) {
            buttonText = 'Connecting...';

        } else if (this.connectionFailed() || this.saveFailed()) {
            buttonText = 'Save changes';
            buttonClick = connect;

            const connectErrorMessage = pathOr(
                null, ['content', 'error'], connectRequest
            );
            const saveErrorMessage = pathOr(
                null, ['content', 'error', 'message'], saveConnectionsRequest
            );
            const genericErrorMessage = 'Hm... had trouble connecting.';
            const errorMessage = connectErrorMessage || saveErrorMessage || genericErrorMessage;
            error = <div className={'errorMessage'}>{errorMessage}</div>;

        } else if (this.isConnected() && this.isSaved()) {
            buttonText = 'Save changes';
            buttonClick = connect;

        } else {
            buttonText = 'Connect';
            buttonClick = connect;
        }

        return (
            <div className={'connectButtonContainer'}>
                <button
                    id="test-connect-button"
                    onClick={buttonClick}
                >
                    {buttonText}
                </button>
                {error}
            </div>
       );
    }

```

shannonlal commented:
@n-riesco I don't think your logic flow worked for the Connect Button.  I will revisit this and step through it to get it working again.  Keep you posted

shannonlal commented:
@n-riesco I think I fixed most of the changes you recommended.  I had to modify the logic flow a little bit for the Button Connect to get it to work but I tested it and it seems to be working.  Please have a look and let me know if you think this makes sense

shannonlal commented:
@n-riesco It looks like there is an issue with the ElasticSearch spec.  I don't think it is related to anything I committed in.  Here is the error:

1) Elasticsearch: connect returns a list of indices:

Would you mind having a look at the error to let me know what might be causing the issue

n-riesco commented on app/components/Settings/ConnectButton/ConnectButton.react.js:
```diff
@@ -39,26 +94,20 @@ export default class ConnectButton extends Component {
-            } else {
+
+        } else if (this.isConnected()) {
+            buttonText = 'Save changes';
```
'`Save changes`' should be shown only after we have successfully connected to a DB (i.e. `this.isConnected()` is `true`) and we have pressed `Edit Credentials` (i.e. `editMode` is `true`).

Also, see that `editMode` is `false` only if we have successfully connected to a DB (i.e. if `editMode` is `false`, then `this.isConnected()` is `true`). This is the reason why I wrote first `if (!editMode`) {...}`.

n-riesco commented:
@shannonlal 

>  I don't think your logic flow worked for the Connect Button. I will revisit this and step through it to get it working again. Keep you posted.

What was the mistake?

n-riesco commented:
@shannonlal 

>  It looks like there is an issue with the ElasticSearch spec. I don't think it is related to anything I committed in. Here is the error:
> 
>     Elasticsearch: connect returns a list of indices:
> 
> Would you mind having a look at the error to let me know what might be causing the issue

The tests are failing because authentication is now enabled in the elasticsearch server.

I've opened PR #382 to disable the tests in CircleCI against the elasticsearch server (we still have the tests against the mocked servers).

n-riesco commented:
Have you seen these comments: [1](https://github.com/plotly/falcon-sql-client/pull/378#discussion_r169026998) and [2](https://github.com/plotly/falcon-sql-client/pull/378#discussion_r169027053)?

Do you prefer instead:?
```
* @param {object} [connectRequest.error]
* @param {string} [connectRequest.error.message] Connection error message
```

shannonlal commented:
@n-riesco I did some more testing and understand how the edit button should be working.  I implemented the Connect Button as you suggested and modified the Jest Test cases to work accordingly.  If you have a chance to look this over I would appreciate it.

n-riesco commented:
@shannonlal LGTM :dancer: 

Please, wait until #382 is merged, so that we can test this PR on CircleCI.

shannonlal commented:
@n-riesco I saw that #382 was just merged.  If tests pass tonight I will merge this PR in.
